File: dqops/src/main/java/com/dqops/metadata/incidents/FilteredNotificationSpec.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */

package com.dqops.metadata.incidents;

import com.dqops.data.incidents.models.IncidentFilteredNotificationLocation;
import com.dqops.metadata.basespecs.AbstractSpec;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.id.HierarchyId;
import com.dqops.metadata.id.HierarchyNodeResultVisitor;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.util.Objects;

/**
 * Notification with filters that is sent only if values in notification message match the filters.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = false)
public class FilteredNotificationSpec extends AbstractSpec {
    private static final ChildHierarchyNodeFieldMapImpl<FilteredNotificationSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractSpec.FIELDS) {
        {
            put("filter", o -> o.filter);
            put("target", o -> o.target);
        }
    };

    @JsonPropertyDescription("Notification filter specification for filtering the incident by the values of its fields.")
    private NotificationFilterSpec filter;

    @JsonPropertyDescription("Notification target addresses for each of the status.")
    private IncidentNotificationTargetSpec target;

    @JsonPropertyDescription("The priority of the notification. Notifications are sent to the first notification targets that matches the filters when processAdditionalFilters is not set.")
    private int priority = 1000;

    @JsonPropertyDescription("Flag to break sending next notifications. Setting to true allows to send next notification from the list in priority order that matches the filter.")
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    private boolean processAdditionalFilters = false;

    @JsonPropertyDescription("Flag to turn off the notification filter.")
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    private boolean disabled = false;

    @JsonPropertyDescription("Message with the details of the filtered notification such as purpose explanation, SLA note, etc.")
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private String message;

    @JsonPropertyDescription("Flag to remove incident that match the filters.")
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    private boolean doNotCreateIncidents = false;

    /**
     * Returns a notification filter.
     * @return Notification filter.
     */
    public NotificationFilterSpec getFilter() {
        return filter;
    }

    /**
     * Sets a notification filter.
     * @param filter Notification filter.
     */
    public void setFilter(NotificationFilterSpec filter) {
        setDirtyIf(!Objects.equals(this.filter, filter));
        this.filter = filter;
        propagateHierarchyIdToField(filter, "filter");
    }

    /**
     * Returns an incident notification target.
     * @return Incident notification target.
     */
    public IncidentNotificationTargetSpec getTarget() {
        return target;
    }

    /**
     * Sets an incident notification target.
     * @param target Incident notification target.
     */
    public void setTarget(IncidentNotificationTargetSpec target) {
        setDirtyIf(!Objects.equals(this.target, target));
        this.target = target;
        propagateHierarchyIdToField(target, "target");
    }

    /**
     * Returns a notification priority.
     * @return Notification priority.
     */
    public int getPriority() {
        return priority;
    }

    /**
     * Sets a notification priority.
     * @param priority Notification priority.
     */
    public void setPriority(int priority) {
        setDirtyIf(!Objects.equals(this.priority, priority));
        this.priority = priority;
    }

    /**
     * Returns a process additional filters flag.
     * @return Process additional filters flag.
     */
    public boolean isProcessAdditionalFilters() {
        return processAdditionalFilters;
    }

    /**
     * Sets a process additional filters flag.
     * @param processAdditionalFilters Process additional filters flag.
     */
    public void setProcessAdditionalFilters(boolean processAdditionalFilters) {
        setDirtyIf(!Objects.equals(this.processAdditionalFilters, processAdditionalFilters));
        this.processAdditionalFilters = processAdditionalFilters;
    }

    /**
     * Returns a disabled flag.
     * @return Disabled flag.
     */
    public boolean isDisabled() {
        return disabled;
    }

    /**
     * Sets a disabled flag.
     * @param disabled Disabled flag.
     */
    public void setDisabled(boolean disabled) {
        setDirtyIf(!Objects.equals(this.disabled, disabled));
        this.disabled = disabled;
    }

    /**
     * Returns a message.
     * @return Message.
     */
    public String getMessage() {
        return message;
    }

    /**
     * Sets a message.
     * @param message Message.
     */
    public void setMessage(String message) {
        setDirtyIf(!Objects.equals(this.message, message));
        this.message = message;
    }

    /**
     * Returns a doNotCreateIncidents flag.
     * @return doNotCreateIncidents flag.
     */
    public boolean isDoNotCreateIncidents() {
        return doNotCreateIncidents;
    }

    /**
     * Sets a doNotCreateIncidents flag.
     * @param doNotCreateIncidents doNotCreateIncidents flag.
     */
    public void setDoNotCreateIncidents(boolean doNotCreateIncidents) {
        setDirtyIf(!Objects.equals(this.doNotCreateIncidents, doNotCreateIncidents));
        this.doNotCreateIncidents = doNotCreateIncidents;
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Calls a visitor (using a visitor design pattern) that returns a result.
     *
     * @param visitor   Visitor instance.
     * @param parameter Additional parameter that will be passed back to the visitor.
     */
    @Override
    public <P, R> R visit(HierarchyNodeResultVisitor<P, R> visitor, P parameter) {
        return visitor.accept(this, parameter);
    }

    /**
     * Creates and returns a copy of this object.
     */
    @Override
    public FilteredNotificationSpec deepClone() {
        FilteredNotificationSpec cloned = (FilteredNotificationSpec) super.deepClone();
        return cloned;
    }

    /**
     * Retrieves the notification name from the hierarchy.
     * @return Notification name or null for a standalone notification spec object.
     */
    @JsonIgnore
    public String getNotificationName() {
        HierarchyId hierarchyId = this.getHierarchyId();
        if (hierarchyId == null) {
            return null;
        }
        return hierarchyId.get(hierarchyId.size() - 1).toString();
    }

    /**
     * Detects where the notification is configured. Is it on the global level, or a connection level.
     * @return Identification of the location where the notification was found, a global level or a connection level.
     */
    @JsonIgnore
    public IncidentFilteredNotificationLocation getNotificationLocation() {
        HierarchyId hierarchyId = this.getHierarchyId();
        if (hierarchyId == null) {
            return null;
        }

        String connectionName = hierarchyId.getConnectionName();
        if (connectionName != null) {
            return IncidentFilteredNotificationLocation.connection;
        }

        return IncidentFilteredNotificationLocation.global;
    }
}

File: distribution/python/dqops/client/models/filtered_notification_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.incident_notification_target_spec import (
        IncidentNotificationTargetSpec,
    )
    from ..models.notification_filter_spec import NotificationFilterSpec


T = TypeVar("T", bound="FilteredNotificationSpec")


@_attrs_define
class FilteredNotificationSpec:
    """
    Attributes:
        filter_ (Union[Unset, NotificationFilterSpec]):
        target (Union[Unset, IncidentNotificationTargetSpec]):
        priority (Union[Unset, int]): The priority of the notification. Notifications are sent to the first notification
            targets that matches the filters when processAdditionalFilters is not set.
        process_additional_filters (Union[Unset, bool]): Flag to break sending next notifications. Setting to true
            allows to send next notification from the list in priority order that matches the filter.
        disabled (Union[Unset, bool]): Flag to turn off the notification filter.
        message (Union[Unset, str]): Message with the details of the filtered notification such as purpose explanation,
            SLA note, etc.
        do_not_create_incidents (Union[Unset, bool]): Flag to remove incident that match the filters.
    """

    filter_: Union[Unset, "NotificationFilterSpec"] = UNSET
    target: Union[Unset, "IncidentNotificationTargetSpec"] = UNSET
    priority: Union[Unset, int] = UNSET
    process_additional_filters: Union[Unset, bool] = UNSET
    disabled: Union[Unset, bool] = UNSET
    message: Union[Unset, str] = UNSET
    do_not_create_incidents: Union[Unset, bool] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        filter_: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.filter_, Unset):
            filter_ = self.filter_.to_dict()

        target: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.target, Unset):
            target = self.target.to_dict()

        priority = self.priority
        process_additional_filters = self.process_additional_filters
        disabled = self.disabled
        message = self.message
        do_not_create_incidents = self.do_not_create_incidents

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if filter_ is not UNSET:
            field_dict["filter"] = filter_
        if target is not UNSET:
            field_dict["target"] = target
        if priority is not UNSET:
            field_dict["priority"] = priority
        if process_additional_filters is not UNSET:
            field_dict["process_additional_filters"] = process_additional_filters
        if disabled is not UNSET:
            field_dict["disabled"] = disabled
        if message is not UNSET:
            field_dict["message"] = message
        if do_not_create_incidents is not UNSET:
            field_dict["do_not_create_incidents"] = do_not_create_incidents

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.incident_notification_target_spec import (
            IncidentNotificationTargetSpec,
        )
        from ..models.notification_filter_spec import NotificationFilterSpec

        d = src_dict.copy()
        _filter_ = d.pop("filter", UNSET)
        filter_: Union[Unset, NotificationFilterSpec]
        if isinstance(_filter_, Unset):
            filter_ = UNSET
        else:
            filter_ = NotificationFilterSpec.from_dict(_filter_)

        _target = d.pop("target", UNSET)
        target: Union[Unset, IncidentNotificationTargetSpec]
        if isinstance(_target, Unset):
            target = UNSET
        else:
            target = IncidentNotificationTargetSpec.from_dict(_target)

        priority = d.pop("priority", UNSET)

        process_additional_filters = d.pop("process_additional_filters", UNSET)

        disabled = d.pop("disabled", UNSET)

        message = d.pop("message", UNSET)

        do_not_create_incidents = d.pop("do_not_create_incidents", UNSET)

        filtered_notification_spec = cls(
            filter_=filter_,
            target=target,
            priority=priority,
            process_additional_filters=process_additional_filters,
            disabled=disabled,
            message=message,
            do_not_create_incidents=do_not_create_incidents,
        )

        filtered_notification_spec.additional_properties = d
        return filtered_notification_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
